design/anoncreds.md

```diff
@@ -0,0 +1,507 @@
+    * Created for each new RevocRegDef.
+    * Used to update RevocReg.
+    * This is different from RevocRegDef Issuer DID (DID used to send `REVOC_REG_DEF` txn) which can be the same for 
+    any number of RevocRegDef.
+* Each `UUID` can be considered as a global UUID within the Ledger.
+The Ledger can guarantee that malicious party can not change/break existing entity 
+defined by a UUID by 
+and checking if there is an entity with the given UUID already existent, plus
+checking the ownership when modifying existing entities (only the issuer who created en entity with the given UUID can modify it). 
+
+      
+### How Prover and Verifier get keys for Credentials and Proofs
+* Proofs and credentials come with `schemaUUID`, `credDefUUID`, `revocDefUUID`.
+* Also there can be `issuanceTime` attribute in each credential (which can be disclosed in the proof).
+* Prover/Verifier looks up SCHEMA using `GET_SCHEMA(schemaUUID)` request to the ledger
+* Prover/Verifier looks up CRED_DEF using `GET_CRED_DEF(credDefUUID, issuanceTime)` request to the ledger
+* Prover/Verifier looks up REVOC_REG_DEF using `GET_REVOC_REG_DEF(revocDefUUID, issuanceTime)` request to the ledger
+* Prover looks up REVOC_REG to update the witness for the given `timestamp` 
+ using `GET_REVOC_REG(revocDefUUID, timestamp)` request to the ledger
+* Verifies looks up REVOC_REG to get accumulator value for the given `timestamp`  
+ using `GET_REVOC_REG(revocDefUUID, timestamp)` request to the ledger
+* Prover and Verifier should look up REVOC_REG by the same `timestamp` when generating and verifying the proof.
+
+### Timestamp Support in State
+
+We need to have a generic way to get the State at the given time.
+- State Trie allows to go to the past, that is given the root hash, get the state with this root.
+- We may have a mapping of each State update (timestamp) to the corresponding root.
+- We need to find a data structure that can help us to find the nearest state timestamp (to get the root) for the given
+time.
+- So, we will be able to get the data (state) at the given time.
+
+This approach can be used for
+* getting `REVOC_REG` at desired time (the same for both proper and verifier),
+possibly long ago in the past;
+* dealing with Requirement 6. 
+
+### Changes in Anoncreds Protocol
+
+If want to support Requirement 6, then the following changes are required in the
+anoncerds protocol:
+
+* Each Credential must have a reserved mandatory attribute: `issuanceTime`.
+    * It's set by the Issuer to specify the time of Issuance.
+    * It's needed to fulfill Requirements 5.
+* This attribute can be considered as `m3` special attribute (`m1` is master secret, `m2` is credential context, `m3` is issuance time).
+* Since the Issuer may be malicious (if keys were compromised already), then 
+a proof that `issuanceTime` is really the current time and not the time from the past is needed.
+    * We can use our blockchain to prepare such a proof.
```

Comment:
> Rather than using the process which engages multiple nodes for each claim issuance, cant the prover just reject the credential when he receives it if issuanceTime is not acceptable.
But what if Prover and Issuer are malicious and Prover will not reject it?
> cant the issuer just embed Sovrin's latest BLS signed state since that contains timestamp too
What prevent Issuer from embeding an old state for the time in past he wants?